core/node/eth_watch/src/event_processors/appended_chain_batch_root_signature.rs

```diff
@@ -0,0 +1,219 @@
+    client::EthClient,
+    event_processors::{EventProcessor, EventProcessorError, EventsSource},
+};
+
+/// Responsible for `AppendedChainBatchRoot` events and saving `BatchAndChainMerklePath` for batches.
+#[derive(Debug)]
+pub struct BatchRootProcessor {
+    next_expected_batch_number: L1BatchNumber,
+    appended_chain_batch_root_signature: H256,
+    merkle_tree: MiniMerkleTree<[u8; 96]>,
+    l2_chain_id: L2ChainId,
+}
+
+impl BatchRootProcessor {
+    pub fn new(
+        next_expected_batch_number: L1BatchNumber,
+        merkle_tree: MiniMerkleTree<[u8; 96]>,
+        l2_chain_id: L2ChainId,
+    ) -> Self {
+        Self {
+            next_expected_batch_number,
+            appended_chain_batch_root_signature: ethabi::long_signature(
+                "AppendedChainBatchRoot",
+                &[
+                    ethabi::ParamType::Uint(256),
+                    ethabi::ParamType::Uint(256),
+                    ethabi::ParamType::FixedBytes(32),
+                ],
+            ),
+            merkle_tree,
+            l2_chain_id,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventProcessor for BatchRootProcessor {
+    async fn process_events(
+        &mut self,
+        storage: &mut Connection<'_, Core>,
+        sl_client: &dyn EthClient,
+        events: Vec<Log>,
+    ) -> Result<usize, EventProcessorError> {
+        let events_count = events.len();
+        let mut transaction = storage
+            .start_transaction()
+            .await
+            .map_err(DalError::generalize)?;
+
+        let grouped_events: Vec<_> = events
+            .into_iter()
+            .map(|log| {
+                let sl_l1_batch_number = L1BatchNumber(
+                    log.l1_batch_number
+                        .expect("Missing L1 batch number for finalized event")
+                        .as_u32(),
+                );
+                let chain_l1_batch_number = L1BatchNumber(h256_to_u256(log.topics[2]).as_u32());
+                let logs_root_hash = H256::from_slice(&log.data.0);
+
+                (sl_l1_batch_number, chain_l1_batch_number, logs_root_hash)
+            })
+            .group_by(|(sl_l1_batch_number, _, _)| *sl_l1_batch_number)
+            .into_iter()
+            .map(|(sl_l1_batch_number, group)| {
+                let group: Vec<_> = group
+                    .into_iter()
+                    .map(|(_, chain_l1_batch_number, logs_root_hash)| {
+                        (chain_l1_batch_number, logs_root_hash)
+                    })
+                    .collect();
+
+                (sl_l1_batch_number, group)
+            })
+            .collect();
+
+        let new_events = grouped_events
+            .into_iter()
+            .skip_while(|(_sl_l1_batch_number, events)| {
+                let first_event = events.first().unwrap();
+                let last_event = events.first().unwrap();
+
+                match (
+                    first_event.0 < self.next_expected_batch_number,
```

Comment:
is this value `self.next_expected_batch_number` ever modified?

Reply:
fixed